mps-sync-plugin/src/main/kotlin/org/modelix/mps/sync/plugin/gui/ModelSyncGuiFactory.kt

```diff
@@ -271,59 +274,78 @@ class ModelSyncGuiFactory : ToolWindowFactory, Disposable {
-                        populateModuleCB()
-                    }
+                val client = existingConnectionsModel.selectedItem as ModelClientV2
+                val repositoryId = repoModel.selectedItem as RepositoryId
+                val branches = client.listBranches(repositoryId)
+
+                branchModel.removeAllElements()
+                branchModel.addAll(branches)
+                if (branchModel.size > 0) {
+                    branchModel.selectedItem = branchModel.getElementAt(0)
+                    populateModuleCB()
                 }
             }
         }
 
-        private fun populateModuleCB() {
+        private suspend fun populateModuleCB() {
             if (existingConnectionsModel.size != 0 && repoModel.size != 0 && branchModel.size != 0) {
-                CoroutineScope(Dispatchers.Default).launch {
-                    val branch =
-                        (existingConnectionsModel.selectedItem as ModelClientV2).getReplicatedModel(branchModel.selectedItem as BranchReference)
-                            .start()
-                    branch.runRead {
-                        moduleModel.removeAllElements()
-                        val children = branch.getRootNode().allChildren.map {
-                            val name = it.getPropertyValue(BuiltinLanguages.jetbrains_mps_lang_core.INamedConcept.name)
-                                ?: it.toString()
-                            INodeWithName(it, name)
-                        }
-                        moduleModel.addAll(children.toList())
-                    }
-                    if (moduleModel.size > 0) {
-                        moduleModel.selectedItem = moduleModel.getElementAt(0)
+                val client = existingConnectionsModel.selectedItem as ModelClientV2
+                val branchReference = branchModel.selectedItem as BranchReference
+
+                val moduleNodes = client.query(branchReference) {
+                    it.allChildren().ofConcept(BuiltinLanguages.MPSRepositoryConcepts.Module).toList()
+                }.map {
+                    val name = it.getPropertyValue(BuiltinLanguages.jetbrains_mps_lang_core.INamedConcept.name)
+                        ?: it.toString()
+                    val id = it.getPropertyValue(BuiltinLanguages.MPSRepositoryConcepts.Module.id) ?: ""
+                    ModuleIdWithName(id, name)
+                }
+
+                moduleModel.removeAllElements()
+                moduleModel.addAll(moduleNodes.toList())
+                if (moduleModel.size > 0) {
+                    moduleModel.selectedItem = moduleModel.getElementAt(0)
+                }
+            }
+        }
+
+        @Synchronized
+        private fun callOnlyIfNotFetching(action: suspend () -> Unit) {
+            if (!isFetchingModulesListFromServer.get()) {
+                isFetchingModulesListFromServer.set(true)
+
+                coroutineScope.launch {
```

Comment:
fixed in c13ac8febff9173492f2f6a3a18e137fd4847849

Reply:
You can use `kotlinx.coroutines.sync.Mutex` instead of `kotlinx.coroutines.sync.Semaphore`.

> Semaphore with permits = 1 is essentially a [Mutex](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.sync/-mutex/index.html).

See https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.sync/-semaphore/

```kt
        private val mutex = Mutex()

        private fun callOnlyIfNotFetching(action: suspend () -> Unit) {
            coroutineScope.launch {
                if (mutex.tryLock()) {
                    try {
                        setUiControlsEnabled(false)
                        action()
                    } catch (ex: Exception) {
                        logger.error(ex) { "Failed to fetch data from server." }
                    } finally {
                        setUiControlsEnabled(true)
                        mutex.unlock()
                    }
                }
            }
        }
 ```

Reply:
This gave me exceptions that when we ewere calling mutex.unlock() then the thread was not in the monitor of the mutex. I.e. we called unlock from a different thread, than where we called tryLock(). However, I'll give it a try again and will let you know in a couple of minutes.

Reply:
Never mind, that was a ReentrantLock instead of a Mutex. Now, it works fine with the Mutex: fefec270781be7272aaad6325b6cad6d15874bce